Refactor `StoreStatsV4PeriodViewController` with a view model

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #5743 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

While working on Home Screen M2 updates with new conversion stats, I started feeling uncomfortable about not having any unit tests on these changes. Store stats are pretty old and we've been adding logic without being able to test them. I decided to take some time to create a view model and move most of the logic that provides view data from the view controller to the view model. My apology for the ~700 diff PR, I tried minimizing the changes but there is a bunch of moved code 😓  The unit tests will be in a separate PR due to the size of this PR. While writing unit tests, I also found some outdated dependency (`currentDate`) and added a new dependency `siteID` to the order and site visit stats predicates.

After this PR, we can also ensure other design updates like for the revenue stats https://github.com/woocommerce/woocommerce-ios/issues/5744 and the time range formatting updates.

Major changes include:

- Created `StoreStatsPeriodViewModel` for `StoreStatsV4PeriodViewController`
- The view model has a publicly settable variable `selectedIntervalIndex` that is set when the user selects a time interval in the chart. Upon changes, all dependent data are recalculated and updated
- Moved all stats text value updates to an observable in the view model: `orderStatsText`, `revenueStatsText`, `visitorStatsText`, `conversionStatsText`
- Other than the stats text values, the following updates are also moved to the view model as an observable:
  - `StatsTimeRangeBarViewModel` which shows the time range text in the header
  - Last updated summary text values
  - Whether the chart should be reloaded (when order stats are updated)
- Removed `currentDate` dependency from `StoreStatsPeriodViewController` and its downstream
- Moved all `ResultsController` storage-level logic from the view controller to the view model, and added `siteID` to the predicates
- Renamed `reloadSiteFields` to `reloadSiteVisitUI`

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Please make sure the order stats function as before:

- Launch the app
- In the My store tab, tap each time range tab --> the stats values and charts should look as before
- Feel free to try switching to dark/light mode

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

jaclync committed:
Move data to view presentation logic to `StoreStatsPeriodViewModel`.

jaclync committed:
Cleanup before code review.

jaclync committed:
Update comments.

jaclync committed:
Delete view model tests file due to large diffs already.

jaclync committed:
Refactor visitor and conversion stats with `assign`.

jaclync committed:
Refactor data observation with `assign` and aggregate order stats and intervals.

jaclync committed:
Minor polishes.

jaclync committed:
Refactor view model public interface to return `AnyPublisher` for observable data.

jaclync committed:
Remove unused dependencies and DI site ID for order and site visit stats predicates.

jaclync committed:
Merge branch 'feat/5743-stats-header-ui' into feat/5743-refactor-stats-view-model

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsPeriodViewModel.swift:
```diff
@@ -0,0 +1,309 @@
+        let siteStatsItems = siteStats?.items?.sorted(by: { (lhs, rhs) -> Bool in
+            return lhs.period < rhs.period
+        }) ?? []
+        if let selectedIndex = selectedIndex, selectedIndex < siteStatsItems.count {
+            guard selectedIndex < siteStatsItems.count else {
+                return nil
+            }
```
I think the `guard` check here is redundant because the condition has been added in the external `if` check. Should we remove it?

jaclync replied:
Ohh great catch! Removed in https://github.com/woocommerce/woocommerce-ios/pull/5788/commits/3285fc119c0a0dec5459faaae330a9032a743aa5

itsmeichigo reviewed (approved):
The changes look great 👏  Thank you for refactoring the code, everything looks so much clearer now! The issue reloading the time range titles when switching range is gone too 🎉 .

One thing I'm wondering: It seems like the last updated date has been removed in the new design, do you think we should remove it on the chart too? The extra information can be useful, but removing it makes the chart clearer so I'm not sure. It depends on whether merchants find it useful too I believe.

jaclync committed:
Remove redundant guard check from CR review.

jaclync committed:
Replace last updated label with a spacer view.

jaclync commented:
> One thing I'm wondering: It seems like the last updated date has been removed in the new design, do you think we should remove it on the chart too? The extra information can be useful, but removing it makes the chart clearer so I'm not sure. It depends on whether merchants find it useful too I believe.

Wow thanks for the catch @itsmeichigo! I missed this design update, and then I discovered other design changes on the top performers section after looking at the design more closely (created a new issue https://github.com/woocommerce/woocommerce-ios/issues/5793) 😅 . I removed the last updated data and replaced the label with a 25px-tall spacer view in the xib https://github.com/woocommerce/woocommerce-ios/pull/5788/commits/97a80ca5e77b41a4b7502fb369283da0dde4be07. It now looks like:

<img src="https://user-images.githubusercontent.com/1945542/147904303-3f73af92-2283-40a3-8700-e34facd2d597.png" width="300" />

See if you'd like to take a look 🙂 I'll merge the PR after CI passes for the follow-up PR with unit tests!

itsmeichigo commented:
Looks good! Thanks for the update @jaclync 😄

I also noticed just now that the numbers for revenues are underneath the horizontal lines in the design (while currently they are aligned) - but I think this is only nice to have and can be complicated to update if the library doesn't support. Just wanted to point it out just in case.

<img width="734" alt="Screen Shot 2022-01-03 at 13 38 57" src="https://user-images.githubusercontent.com/5533851/147904874-26d49487-fe84-4fcd-baa2-62a9cbe00a3f.png">

jaclync commented:
> I also noticed just now that the numbers for revenues are underneath the horizontal lines in the design (while currently they are aligned) - but I think this is only nice to have and can be complicated to update if the library doesn't support. Just wanted to point it out just in case.

Thank you for spotting this! 😮 🦅 👁️ I just added this to the design polishes checklist in https://github.com/woocommerce/woocommerce-ios/issues/5745 - there might be a way to do this with the Charts framework, I will look into this after the major updates are in.
